lib/tpm2_errata.c

```diff
@@ -0,0 +1,212 @@
+    .name = "Sign/decrypt attribute encoding",
+};
+
+/* Add the new errata descriptor here */
+
+static struct tpm2_errata_desc *errata_desc_list[] = {
+    [SPEC_116_ERRATA_2_7] = &sign_decrypt_attribute_encoding,
+    /* Append the new errata descriptor here */
+};
+
+static bool errata_match(struct tpm2_errata_desc *errata) {
+
+    return errata->spec_rev >= this_errata_info->spec_rev &&
+           errata->errata_ver > this_errata_info->errata_ver;
+}
+
+static bool errata_query(int nr, struct tpm2_errata_desc **errata) {
+
+    if (nr < 0 || (size_t)nr >= sizeof(errata_desc_list) / sizeof(*errata)) {
+        LOG_ERR("Invalid errata requested: %d\n", nr);
+        return false;
+    }
+
+    if (!this_errata_info) {
+        LOG_ERR("Unknow TPM_SPEC\n");
+        return false;
+    }
+
+    *errata = errata_desc_list[nr];
+
+    return errata_match(*errata);
+}
+
+bool tpm2_errata_fixup(int nr, ...) {
+
+    struct tpm2_errata_desc *errata;
+    va_list ap;
+
+    if (errata_query(nr, &errata) == false)
+        return false;
+
+    va_start(ap, nr);
+    errata->fixup(errata, &ap);
+    va_end(ap);
+
+    return true;
+}
+
+const char *tpm2_errata_name(int nr) {
+
+    struct tpm2_errata_desc *errata;
+
+    if (errata_query(nr, &errata) == false)
+        return false;
+
+    return errata->name;
+}
+
+void tpm2_errata_init(TSS2_SYS_CONTEXT *sapi_ctx) {
+
+    TPMS_CAPABILITY_DATA      capability_data;
+    TPML_TAGGED_TPM_PROPERTY *properties;
+    TSS2_RC                   rc;
+    TPMI_YES_NO               more_data;
+    size_t                    i;
+    UINT32                    spec_rev, day_of_year, year;
+
+    rc = Tss2_Sys_GetCapability(sapi_ctx,
+                                NULL,
+                                TPM_CAP_TPM_PROPERTIES,
+                                PT_FIXED,
+                                MAX_TPM_PROPERTIES,
+                                &more_data,
+                                &capability_data,
+                                NULL);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERR("Failed to GetCapability: capability: 0x%x, property: 0x%x, "
+                "TSS2_RC: 0x%x\n", TPM_CAP_TPM_PROPERTIES, PT_FIXED, rc);
+        return;
+    } else if (more_data == YES)
+        LOG_WARN("More data to be queried: capability: 0x%x, property: "
+                 "0x%x\n", TPM_CAP_TPM_PROPERTIES, PT_FIXED);
+
+    spec_rev = day_of_year = year = 0;
+    properties = &capability_data.data.tpmProperties;
+    for (i = 0; i < properties->count; ++i) {
+        TPMS_TAGGED_PROPERTY *property = properties->tpmProperty + i;
+
+        if (property->property == TPM_PT_REVISION)
+            spec_rev = property->value;
+        else if (property->property == TPM_PT_DAY_OF_YEAR)
+            day_of_year = property->value;
+        else if (property->property == TPM_PT_YEAR)
+            year = property->value;
+        else if (property->property > TPM_PT_YEAR)
+            break;
```

Comment:
Probably is worth adding a comment that TPM_PT_YEAR and TPM_PT_DAY_OF_YEAR only have the values of the release date of the specification if the TPM does not implement an errata, but that the values indicate the release date of the errata document if the TPM implements the errata.

This wasn't evident to me and I had to read the TPM library part 2 to figure out.